Support BF16ImmPtr

EikanWang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #85140
* #84402
* __->__ #84041

- To support BF16 Immediate value by converting it to uint16. The behavior is as same as BF16 tensor
- Enable BF16 test cases.

EikanWang committed:
Support BF16ImmPtr

[ghstack-poisoned]

EikanWang committed:
Update on "Support BF16ImmPtr"




[ghstack-poisoned]

EikanWang committed:
Update on "Support BF16ImmPtr"




[ghstack-poisoned]

EikanWang committed:
Update on "Support BF16ImmPtr"




[ghstack-poisoned]

EikanWang committed:
Update on "Support BF16ImmPtr"




[ghstack-poisoned]

EikanWang committed:
Update on "Support BF16ImmPtr"




[ghstack-poisoned]

EikanWang committed:
Update on "Support BF16ImmPtr"




[ghstack-poisoned]

EikanWang committed:
Update on "Support BF16ImmPtr"




[ghstack-poisoned]

EikanWang committed:
Update on "Support BF16ImmPtr"


- To support BF16 Immediate value by converting it to uint16. The behavior is as same as BF16 tensor
- Enable BF16 test cases.

[ghstack-poisoned]

EikanWang committed:
Update on "Support BF16ImmPtr"


- To support BF16 Immediate value by converting it to uint16. The behavior is as same as BF16 tensor
- Enable BF16 test cases.

[ghstack-poisoned]

ZolotukhinM commented on test/test_tensorexpr.py:
```diff
@@ -854,33 +885,45 @@ def test_threshold(x, y):
+            if torch_fn == test_lgamma and dev == "cuda":
+                # lgamma_cuda does not support BF16
+                continue
+            torch.manual_seed(0)
+            torch.set_printoptions(10)
```
This probably can be hoisted from the loop.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/half_support.h:
```diff
@@ -101,7 +101,8 @@ class HalfRewriter : public IRMutator {
     if (isHalf(v)) {
       if (inserted_half_casts_.count(v) < 1) {
-        return child;
+        v->set_dtype(v->dtype().cloneWithScalarType(c10::kFloat));
+        return v;
```
Shouldn't this be `child` instead of `v`? And why do we need this change at all?

EikanWang replied:
The child value could be long or double as follows. Then the immediate long value `2` needs to be converted to bloat16. But we do not support the conversion between non-float32 and bfloat16. Hence, I always update the `Cast` to cast the non-float32 to float32.
```python
a = torch.randn(10).bfloat16()
c = a + 2
```

ZolotukhinM replied:
I think in that case we need to insert an extra cast then? I.e. first convert to f32, then from f32 to bf16. I'm not sure bluntly overwriting dtype is a correct solution here. Also, could you please add tests exercising these type conversions? I believe current tests might not be covering that.

EikanWang replied:
> I think in that case we need to insert an extra cast then? I.e. first convert to f32, then from f32 to bf16.

It is sort of complex. Because the immediate value of the python script will be translated to `LongImm` or `DoubleImm`, and it is hard to know if we need to insert `Cast` because it depends on the `Expression`. Take an example as follows, the `2` need to be cast to `Float` but not `BFloat16`.

```python
a = torch.randn(10)
c = a + 2
```

frank-wei replied:
Just clarify, you will cast non-float32 to float32 first. And then cast to bfloat16 or no change depends on current expression? If we add cast there, the purpose is also casting it to float32. I feel there is no different there.

EikanWang replied:
> you will cast non-float32 to float32 first. And then cast to bfloat16 or no change depends on current expression?

Yes, we always cast the non-float32 to float32 first and then cast to bfloat16 if needed. 

> If we add cast there, the purpose is also casting it to float32. I feel there is no difference there.

If we add cast for each immediate data type, we have to enumerate all the Immediate value types like `LongImm`, `IntImm`, `ByteImm`. So that's why I just change the destination datatype of `cast` to `float`

ZolotukhinM replied:
> If we add cast for each immediate data type, we have to enumerate all the Immediate value types like LongImm, IntImm, ByteImm. So that's why I just change the destination datatype of cast to float

If that's what needs to be done to accurately represent the original behavior, then we should do that. Otherwise we will simply break models when these casts actually happen.

EikanWang replied:
> If that's what needs to be done to accurately represent the original behavior, then we should do that. Otherwise we will simply break models when these casts actually happen.

I'm not sure if the immediate value is the only case. But anyway, I will double-check it.

ZolotukhinM reviewed (approved):
LGTM with a minor remark/question.

EikanWang committed:
Update on "Support BF16ImmPtr"


- To support BF16 Immediate value by converting it to uint16. The behavior is as same as BF16 tensor
- Enable BF16 test cases.

[ghstack-poisoned]

EikanWang committed:
Update on "Support BF16ImmPtr"


- To support BF16 Immediate value by converting it to uint16. The behavior is as same as BF16 tensor
- Enable BF16 test cases.

[ghstack-poisoned]

EikanWang committed:
Update on "Support BF16ImmPtr"


- To support BF16 Immediate value by converting it to uint16. The behavior is as same as BF16 tensor
- Enable BF16 test cases.

[ghstack-poisoned]

EikanWang committed:
Update on "Support BF16ImmPtr"


- To support BF16 Immediate value by converting it to uint16. The behavior is as same as BF16 tensor
- Enable BF16 test cases.

[ghstack-poisoned]

EikanWang committed:
Update on "Support BF16ImmPtr"


- To support BF16 Immediate value by converting it to uint16. The behavior is as same as BF16 tensor
- Enable BF16 test cases.

[ghstack-poisoned]

EikanWang committed:
Update on "Support BF16ImmPtr"


- To support BF16 Immediate value by converting it to uint16. The behavior is as same as BF16 tensor
- Enable BF16 test cases.

[ghstack-poisoned]

EikanWang committed:
Update on "Support BF16ImmPtr"


- To support BF16 Immediate value by converting it to uint16. The behavior is as same as BF16 tensor
- Enable BF16 test cases.

[ghstack-poisoned]

frank-wei commented:
@pytorchbot merge -l

frank-wei commented:
Could you pls check the failed tests?

EikanWang committed:
Update on "Support BF16ImmPtr"


- To support BF16 Immediate value by converting it to uint16. The behavior is as same as BF16 tensor
- Enable BF16 test cases.

[ghstack-poisoned]

EikanWang commented:
> Could you pls check the failed tests?

Yes. I fixed it and triggered the CI again. I will check the CI result.

EikanWang committed:
Update on "Support BF16ImmPtr"


- To support BF16 Immediate value by converting it to uint16. The behavior is as same as BF16 tensor
- Enable BF16 test cases.

[ghstack-poisoned]

EikanWang committed:
Update on "Support BF16ImmPtr"


- To support BF16 Immediate value by converting it to uint16. The behavior is as same as BF16 tensor
- Enable BF16 test cases.

[ghstack-poisoned]

EikanWang commented:
> > Could you pls check the failed tests?
> 
> Yes. I fixed it and triggered the CI again. I will check the CI result.

@frank-wei , this PR passed the CI. Could you review it?

EikanWang commented:
@pytorchbot merge -l

EikanWang committed:
Update on "Support BF16ImmPtr"


- To support BF16 Immediate value by converting it to uint16. The behavior is as same as BF16 tensor
- Enable BF16 test cases.

[ghstack-poisoned]

EikanWang committed:
Update on "Support BF16ImmPtr"


- To support BF16 Immediate value by converting it to uint16. The behavior is as same as BF16 tensor
- Enable BF16 test cases.

[ghstack-poisoned]

EikanWang commented:
@pytorchbot merge -l
